lib/containers/runtime.pm

```diff
@@ -0,0 +1,143 @@
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Summary: Abstraction layer to operate docker and podman containers through same interfaces
+# Maintainer: qac team <qa-c@suse.de>
+
+package containers::runtime;
+use Mojo::Base -base;
+use testapi;
+use Test::Assert 'assert_equals';
+
+has runtime => undef;
+
+sub _rt_assert_script_run {
+    my ($self, $cmd, @args) = @_;
+    assert_script_run($self->runtime . " " . $cmd, @args);
+}
+
+sub _rt_script_run {
+    my ($self, $cmd, @args) = @_;
+    return script_run($self->runtime . " " . $cmd, @args);
+}
+
+sub _rt_script_output {
+    my ($self, $cmd, @args) = @_;
+    return script_output($self->runtime . " " . $cmd, @args);
+}
+
+=head2 build
+
+Build a container after visit C<dockerfile_path>.
+C<dockerfile_path> is the directory with the Dockerfile.
+C<container_tag> will be the name of the container.
+
+=cut
+sub build {
+    my ($self, $dockerfile_path, $container_tag) = @_;
+    die 'wrong number of arguments' if @_ < 3;
+    assert_script_run "cd $dockerfile_path";
+    $self->_rt_assert_script_run("build -t $container_tag .");
+    record_info "$container_tag created";
+}
+
+=head2 up
+
+Run a container.
+C<image_name> is required and can be the image id, the name or name with tag.
+If C<daemon> is enabled then container will run in the detached mode. Otherwise will be in the 
+interactive mode.
+if C<cmd> found then it will execute the given command into the container.
+The container is always removed after exit.
+
+=cut
+sub up {
+    my ($self, $image_name, %args) = @_;
+    die 'image name or id is required' unless $image_name;
+    my $mode   = $args{daemon} ? '-d'    : '-it';
+    my $remote = $args{cmd}    ? 'sh -c' : '';
+    my $ret    = $self->_rt_script_run(sprintf qq(run --rm %s %s %s '%s'), $mode, $image_name, $remote, $args{cmd});
+    record_info "Remote run on $image_name", $args{cmd};
+    return $ret;
+}
+
+=head2 pull
+
+Pull a container with the given C<tag>.
+
+=cut
+sub pull {
+    my ($self, $tag) = @_;
+    my $ret = $self->_rt_script_run("pull $tag");
+    return $ret;
+}
+
+=head2 enum_images
+
+Return an array ref of the images
+
+=cut
+sub enum_images {
+    my ($self) = shift;
+    my $images_s = $self->_rt_script_output("images -q");
+    record_info "Images", $images_s;
+    my @images = split /[\n\t]/, $images_s;
+    return \@images;
+}
+
+=head2 enum_images
+
+Return an array ref of the containers
+
+=cut
+sub enum_containers {
+    my ($self) = shift;
+    my $containers_s = $self->_rt_script_output("container ls -q");
+    record_info "Containers", $containers_s;
+    my @containers = split /[\n\t]/, $containers_s;
+    return \@containers;
+}
+
+=head2 info
+
+Assert a C<property> against given expected C<value>.
+
+=cut
+sub info {
```

Comment:
```suggestion
sub assert_info {
```
or anything else which give more context about what actually happening here

Reply:
The idea is that when you run `docker info` you expect an output to validate a property. but you point me to another problem in this function. So i will keep the name and let me check the other problem.

Reply:
I would not expect from a function with name `info` to assert something

Reply:
Sorry. no problem. The name is fine and i can explain. I update the perlpod for i hope it makes it easier to understand. the main purpose is to get the `info` data. if you want to validate a specific property you can pass a second argument and then the method greps this. the assert_script_run ensures that the command was successful if the variable is the expected otherwise fails.

Reply:
I have to admit that single purpose functions are cooler while they are keeping the spirit of [The Unix Tools Philosophy](https://linux.die.net/Linux-CLI/c1089.htm).